src/libmongoc/src/mongoc/mongoc-stream-tls-openssl.c

```diff
@@ -853,6 +853,151 @@ mongoc_stream_tls_openssl_new (mongoc_stream_t *base_stream, const char *host, m
    RETURN ((mongoc_stream_t *) tls);
 }
 
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+/*
+ *--------------------------------------------------------------------------
+ *
+ * mongoc_stream_tls_openssl_new_with_context --
+ *
+ *       Creates a new mongoc_stream_tls_openssl_t to communicate with a remote
+ *       server using a TLS stream, using an existing OpenSSL context.
+ *
+ *       Only called by mongoc_stream_tls_new_with_hostname_and_openssl_context.
+ *
+ *       @ssl_ctx is the global OpenSSL context for the mongoc_client_t
+ *       associated with this function call.
+ *
+ * Returns:
+ *       NULL on failure, otherwise a mongoc_stream_t.
+ *
+ * Side effects:
+ *       None.
+ *
+ *--------------------------------------------------------------------------
+ */
+
+mongoc_stream_t *
+mongoc_stream_tls_openssl_new_with_context (
```

Comment:
Suggest having `mongoc_stream_tls_openssl_new_with_context` and `mongoc_stream_tls_openssl_new` pass the `SSL_CTX` to a shared implementation:

```c
// `create_stream_with_openssl_ctx` creates a stream with `ssl_ctx`.
static mongoc_stream_t *
create_stream_with_openssl_ctx (
   mongoc_stream_t *base_stream, const char *host, mongoc_ssl_opt_t *opt, int client, SSL_CTX *ssl_ctx)
{
   // TODO.
}

// `mongoc_stream_tls_openssl_new` creates a stream with a new OpenSSL context.
mongoc_stream_t *
mongoc_stream_tls_openssl_new (mongoc_stream_t *base_stream, const char *host, mongoc_ssl_opt_t *opt, int client)
{
   SSL_CTX *ssl_ctx = _mongoc_openssl_ctx_new (opt);

   if (!ssl_ctx) {
      return NULL;
   }

   return create_stream_with_openssl_ctx (base_stream, host, opt, client, ssl_ctx); // Takes ownership of `ssl_ctx`.
}

#if OPENSSL_VERSION_NUMBER >= 0x10100000L
// `mongoc_stream_tls_openssl_new_with_context` creates a stream from an existing OpenSSL context.
mongoc_stream_t *
mongoc_stream_tls_openssl_new_with_context (
   mongoc_stream_t *base_stream, const char *host, mongoc_ssl_opt_t *opt, int client, SSL_CTX *ssl_ctx)
{
   BSON_ASSERT_PARAM (ssl_ctx);
   SSL_CTX_up_ref (ssl_ctx);                                                        // Increment reference counter.
   return create_stream_with_openssl_ctx (base_stream, host, opt, client, ssl_ctx); // Takes ownership of reference.
}
#endif
```